src/irmin-pack/s.ml

```diff
@@ -55,41 +55,92 @@ module type Specifics = sig
+        useful for debugging. The default is true.
 
-      TODO: Detail exceptions raised. *)
+        If [throttle] is [Skip] and there is a concurrent gc, [start_exn]
+        returns false immediately without launching a second gc. If [throttle]
+        is [Block] and there is a concurrent gc, [start_exn] blocks waiting for
+        the previous gc to finish and only after launches a second gc. If the
+        previous GC failed, the function returns without launching a new GC.
 
-  val finalise_gc : ?wait:bool -> repo -> bool Lwt.t
-  (** [finalise_gc ?wait repo] waits for the gc process to finish in order to
-      finalise it. It returns true if a GC was finalised.
+        TODO: Detail exceptions raised. *)
 
-      Finalising consists of mutating [repo] so that it points to the new file
-      and to flush the internal caches that could be referencing GCed objects.
+    val finalise_exn : ?wait:bool -> repo -> bool Lwt.t
+    (** [finalise_exn ?wait repo] waits for the gc process to finish in order to
+        finalise it. It returns true if a GC was finalised.
 
-      If [wait = true] (the default), the call blocks until the GC process
-      finishes. If [wait = false] it either returns [false], raises an exception
-      or finalises and returns [true].
+        Finalising consists of mutating [repo] so that it points to the new file
+        and to flush the internal caches that could be referencing GCed objects.
 
-      If there are no running gcs, the call is a no-op and it returns false.
+        If [wait = true] (the default), the call blocks until the GC process
+        finishes. If [wait = false] it either returns [false], raises an
+        exception or finalises and returns [true].
+
+        If there are no running gcs, the call is a no-op and it returns false.
+
+        TODO: Detail exceptions raised. *)
+
+    (** {2 High-level API} *)
+
+    val run :
+      ?finished:(float -> unit) ->
+      repo ->
+      commit_key ->
+      (bool, string) result Lwt.t
+    (** [run repo commit_key] attempts to start a GC process for a [repo] by
+        discarding all data prior to [commit_key]. If a GC process is already
+        running, a new one will not be started.
 
-      TODO: Detail exceptions raised. *)
+        [run] will also finalise the GC process automaticlly. For more detailed
+        control, see {!start_exn} and {!finalise_exn}.
+
+        When the GC process is finalised, [finished] is called with the time
+        elapsed.
+
+        To monitor progress of GC, see {!wait} or {!is_finished}.
+
+        Returns whether a GC process successfully started or not.
+
+        All exceptions that [Irmin_pack] knows how to handle are caught and
+        returned as pretty-print strings; others are re-raised. The error string
+        should be used only for informational purposes, like logging. *)
+
+    val wait : repo -> (unit, string) result Lwt.t
+    (** [wait repo] blocks until GC is finished or is idle.
```

Comment:
what happens if a concurrent `run` is failing? Will both callers get reported the same error?

Reply:
Good question. For the unix errors, I don't know off the top of my head the behavior of `waitpid`. Does it return the same information to multiple invocations? My guess is yes, but I don't know. For reading the gc results, I would expect them to receive the same information based on whatever is saved on disk as the gc result.